Adding the first iteration of container support (for unit testing)

johlju opened the pull request:
- Add opt-in parameter RunInContainer for the helper function Invoke-AppveyorTestScriptTask
  which enables running unit tests in a Docker Windows container. The RunInContainer
  parameter can only be used when the parameter RunTestInOrder is also used.
- Moved helper function Write-Info to the TestHelpers module so it could be reused
  by other modules. The string was changed to output 'UTC' before the time to
  clarify that it is UTC time, and optional parameter 'ForegroundColor' was added
  so that it is possible to change the color of the text that is written.
- Added module DscResource.Container which contain logic to handle the container
  testing when unit tests are run in a Docker Windows container.

**This PR will speed up testing in SqlServerDsc with ~20 minutes.**

This PR only support running container test on all unit tests (in \Tests\Unit folder), and only when both parameters `-RunInContainer` and `-RunTestInOrder` are used.
I'm planning of adding support for multiple containers and the option to choose which test runs in which container.  See issue #204. My thought is that issue #204 remains open until next iteration is finished. 

This is tested with SharePointDsc and PSDscResources to make sure the changes does not change any existing functionality.
**SharePointDsc:** https://ci.appveyor.com/project/johlju/sharepointdsc-p26e6
**PSDscResources:** https://ci.appveyor.com/project/johlju/psdscresources

This is tested using the docker functionality here with SqlServerDsc.
**SqlServerDsc:** https://ci.appveyor.com/project/johlju/sqlserverdsc

johlju commented:
@PlagueHO do you have some time to review this one? No hurry. Maybe we should merge this one after the next release. Also, I'm interested if you see anything that could be improved.

PlagueHO commented:
@johlju - utterly fantastic work here! This is super exciting and I'm looking forward to possibly use this in xNetworking (the tests are starting to get loooong). Have you thought much on how we might implement this in the harness type resources? I might do some experimenting into using this once it has been merged.

Anyway, again, well done and great stuff.

---

Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, 25 unresolved discussions, some commit checks failed.

---

*[AppVeyor.psm1, line 127 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4SUVUxszok8avm11By:-L4SUVUxszok8avm11Bz:b-jyiqx0) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/AppVeyor.psm1#L127)):*
> ```PowerShell
> 
>     .PARAMETER RunTestInOrder
>         This will run the integration tests in order. First common tests will run,
> ```

Some minor grammatical tweaks (non-blocking):
This will cause the integration tests to be run in order. First, the common tests will run, followed by the unit tests. Finally the integration tests will be run in the order defined. Each integration test...

---

*[AppVeyor.psm1, line 131 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4SUNgpOTMUuA0Y4IrZ:-L4SUNgpOTMUuA0Y4Ir_:b-tm5z8m) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/AppVeyor.psm1#L131)):*
> ```PowerShell
>         Each integration test configuration file ('*.config.ps1') must be decorated
>         with an attribute `Microsoft.DscResourceKit.IntegrationTest` containing
>         a named attribute argument 'OrderNumber' and be assigned a numeric value
> ```

Might be worth making it clear what will happen if an integration test is not decorated (e.g. does it run first or last?)

---

*[AppVeyor.psm1, line 382 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4SVWVu_Ex62NdJRCjB:-L4SVWVv4YeS6nbXaQTY:b-5rl7sr) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/AppVeyor.psm1#L382)):*
> ```PowerShell
>                     - Secondly the tests that use mocks will run (unit tests),
>                       unless they should be run in a container.
>                     - Thirdly, those the tests that actually changes things
> ```

Thirdly -> Finally

---

*[AppVeyor.psm1, line 412 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4T6LLx2KdpVaOumZ0N:-L4T6LLx2KdpVaOumZ0O:bn9ztea) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/AppVeyor.psm1#L412)):*
> microsoft/windowsservercore
Is it worth specifying a tag? E.g. 'microsoft/windowsservercore:latest'

Otherwise it might quite happily use old/out-of-date docker image versions if they exist on the host (I'm often guilty of not updating my docker images :grin: ). 

Non-blocking BTW - just a thought.

---

*[AppVeyor.psm1, line 420 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4SXxgPqtfjlpy83L6H:-L4SXxgQVV32khUtluN_:b-runlzs) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/AppVeyor.psm1#L420)):*
> ```PowerShell
>                         If code coverage was chosen, then evaluate the files
>                         needed to be able to calculate coverage for the files
>                         tested in the container. The rest of the files is left
> ```

'The rest of the files is left' -> 'The rest of the files are left'

---

*[AppVeyor.psm1, line 476 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4SZ1J3HDhEXq_ik381:-L4SZ1J3HDhEXq_ik382:by6qzl7) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/AppVeyor.psm1#L476)):*
> ```PowerShell
> 
>                     $containerIdentifier = New-Container @newContainerParameters
>                     Start-Container -ContainerIdentifier $containerIdentifier | Out-Null
> ```

Is it possible that the container won't start for whatever reason?

---

*[AppVeyor.psm1, line 535 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4SZQlJsitx0ANEzy-G:-L4SZQlJsitx0ANEzy-H:b-7etzjl) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/AppVeyor.psm1#L535)):*
> ```PowerShell
>                 {
>                     $containerExitCode = Wait-Container -ContainerIdentifier $containerIdentifier
>                     if ($containerExitCode -ne 0)
> ```

Blank line before this line?

---

*[README.md, line 399 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4SbvwfvozBfiwFHo5x:-L4SbvwgpzDLyLXuKmqu:b99itgi) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/README.md#L399)):*
> ```Markdown
> tests in parallell.
> 
> The Pester output from the container, including errors will be outputted
> ```

non-blocking nitpick: 'outputted on the console` -> `sent to the console`

---

*[README.md, line 409 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4ScINzZhjCt8RtATTj:-L4ScINzZhjCt8RtATTk:b-xnglto) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/README.md#L409)):*
> ```Markdown
> 
> If the container returns an exit code other than 0, the Docker log for the
> container is gathered and uploaded as an artifact. This to hopefully be able
> ```

Non-blocking grammatical suggestion: 'This to hopefully be able to see more detailed error of what happened' -> 'This is intended to enable a more detailed error of what happened to be displayed'

---

*[README.md, line 411 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4ScfHe4Z-FAoqa7SWN:-L4ScfHe4Z-FAoqa7SWO:bnkf1kg) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/README.md#L411)):*
> ```Markdown
> container is gathered and uploaded as an artifact. This to hopefully be able
> to see more detailed error of what happened.
> From the Docker log an error record is searched, and if found and an error will
> ```

Non-blocking grammar: 'From the Docker log an error record is searched, and if found and an error will be thrown stopping the tests in the build worker' -> 'The Docker log will be searched for any error records. If any are found then an exception will be thrown which will stop the the tests in the build worker'. 

---

*[README.md, line 425 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4SdTSc8CAAd9t5r4E5:-L4SdTSdLbVVfGUe-Cb-:b-2py3s7) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/README.md#L425)):*
> ```Markdown
>   returned after it finished the test run in the container.
> * unittest_DockerLog.txt - If the container exits with any other exit code
>   than 0 a Docker log is gathered and uploaded as an artifact. This to
> ```

Non-blocking grammar: `This is intended to enable a more detailed view of the error.`

---

*[README.md, line 427 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4SdEcIFtCsPs4G8YrU:-L4SdEcIFtCsPs4G8YrV:b-pddat6) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/README.md#L427)):*
> ```Markdown
>   than 0 a Docker log is gathered and uploaded as an artifact. This to
>   hopefully be able to see more detailed error of what happened.
>   If the container exits with exit code 0 no Docker log is uploaded.
> ```

Non-blocking grammar: If the container exits with exit code 0 then the Docker log will _not_ be uploaded.

---

*[TestHelper.psm1, line 1456 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4SeZmda8jyiX8LOUDx:-L4SeZmda8jyiX8LOUDy:b-xnm5) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/TestHelper.psm1#L1456)):*
> ```PowerShell
> 
> Export-ModuleMember -Function @(
>     'New-Nuspec', `
> ```

Nitpick: Could we use back-tick on all lines or on no lines for consistency? My preference in this case is no back-tick.

---

*[DscResource.Container/DscResource.Container.psm1, line 9 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4T-OBQb26uLzusFIBH:-L4T-OBQb26uLzusFIBI:b-pnee1z) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/DscResource.Container/DscResource.Container.psm1#L9)):*
> ```PowerShell
> <#
>     .SYNOPSIS
>         This is the function that is run the tests in the container.
> ```

Could we clarify that this is intended to be executed within the container itself? E.g.

This is the function that is executed inside the container to run the tests.

---

*[DscResource.Container/DscResource.Container.psm1, line 101 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4SiXBIr_YGE5WTFH5q:-L4SiXBIr_YGE5WTFH5r:bk2iax6) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/DscResource.Container/DscResource.Container.psm1#L101)):*
> ```PowerShell
> <#
>     .SYNOPSIS
>         Creates a new container a prepares it to run tests.
> ```

'Creates a new container a prepares it to run tests.' -> 'Creates a new container and prepares it to run tests.'

---

*[DscResource.Container/DscResource.Container.psm1, line 135 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4SihDxvog3fBLhT9hM:-L4SihDyQmW-ZEOjX-zA:b-4cxzef) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/DscResource.Container/DscResource.Container.psm1#L135)):*
> ```PowerShell
>         [Parameter(Mandatory = $true)]
>         [System.String]
>         $ImageName,
> ```

Non-blocking nitpick: Could we default this value to simplify calls? In general I'd expect we'll always use `windowsservercore:latest` - is that correct?

---

*[DscResource.Container/DscResource.Container.psm1, line 310 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4T0EWJghkLg3NdkHLT:-L4T0EWJghkLg3NdkHLU:bh5ru7u) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/DscResource.Container/DscResource.Container.psm1#L310)):*
> ```PowerShell
> <#
>     .SYNOPSIS
>         Waits for a container to stopp (exit).
> ```

stopp -> stop

---

*[DscResource.Container/DscResource.Container.psm1, line 346 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4T0QkElRrFjSH9bXx_:-L4T0QkElRrFjSH9bXxa:b43buj4) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/DscResource.Container/DscResource.Container.psm1#L346)):*
> ```PowerShell
>             # This returns a string with either 'false' or 'true'.
>             $containerRunning = docker inspect $ContainerIdentifier --format '{{.State.Running}}'
>         } until ($containerRunning -eq 'false')
> ```

Non-blocking: Is it worth implementing some sort of timeout here? Or perhaps this could be something looked at in a later version once we've had a chance to use this in anger a bit more.

---

*[DscResource.Container/DscResource.Container.psm1, line 539 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4T1snMf242E0xpA6JM:-L4T1snNAez1llroLzO2:bjuzu6d) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/DscResource.Container/DscResource.Container.psm1#L539)):*
> ```PowerShell
>             } | Select-Object -ExpandProperty 'Context' -Unique
> 
>             foreach ($contextBlockName in $uniqueContextBlockName)
> ```

Will this support Pester nested Context blocks? Could add it to the future TODO list if not supported as some modules have used this. Probably not worth doing now though. So non-blocking. Could add this to the .NOTES for the function though.

---

*[DscResource.Container/DscResource.Container.psm1, line 682 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4T33Gx-bXRE1WlbOlp:-L4T33GyZdTF-y2T7ao6:brxcq2z) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/DscResource.Container/DscResource.Container.psm1#L682)):*
> ```PowerShell
> 
>         [PSCustomObject[]] $MissedCommand = $MissedCommand |
>             Select-Object @{
> ```

Should be named parameter here.

---

*[Tests/Unit/DscResource.Container.Tests.ps1, line 153 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4T4VufT6vfMVRJCA4v:-L4T4VugtJ1aRIgB-7EV:bopobba) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/Tests/Unit/DscResource.Container.Tests.ps1#L153)):*
> ```PowerShell
>             }
> 
>             It 'Should call the correct mocks for the Describe-block' {
> ```

Nitpick: add 'without throwing'

---

*[Tests/Unit/DscResource.Container.Tests.ps1, line 176 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4T4dScdr_UdUPT529j:-L4T4dScdr_UdUPT529k:bopobba) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/Tests/Unit/DscResource.Container.Tests.ps1#L176)):*
> ```PowerShell
>             }
> 
>             It 'Should call the correct mocks for the Describe-bloc' {
> ```

Nitpick: add 'without throwing'

---

*[Tests/Unit/DscResource.Container.Tests.ps1, line 335 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4T5dRIle7dWeAGRo06:-L4T5dRIle7dWeAGRo07:by75bwc) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/Tests/Unit/DscResource.Container.Tests.ps1#L335)):*
> ```PowerShell
>             <#
>                 A placeholder/wrapper for the docker.exe so the code is tricked
>                 in thinking it exist, and so that we can mock it.
> ```

"in to thinking it exists so that we can mock it."

And throughout.

---

*[Tests/Unit/DscResource.Container.Tests.ps1, line 345 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4T4nOzw9aOHj2PuMQo:-L4T4nOzw9aOHj2PuMQp:bl13k5z) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/Tests/Unit/DscResource.Container.Tests.ps1#L345)):*
> ```PowerShell
> 
>             Mock -CommandName 'docker' -ParameterFilter $mockDockerCpTo_ParameterFilter
> ```

Remove blank line

---

*[Tests/Unit/DscResource.Container.Tests.ps1, line 550 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4T5IQrgo0jBgYfEZZe:-L4T5IQrgo0jBgYfEZZf:bvzs3gz) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/Tests/Unit/DscResource.Container.Tests.ps1#L550)):*
> ```PowerShell
>             }
> 
>             It 'Should wait for the container until it is stoppen without throwing' {
> ```

SWE :grin: stoppen -> stopped

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-:-L4T6i2qaxHDjoHSfb7p:bopd5da)*
<!-- Sent from Reviewable.io -->

johlju commented:
Thanks! It to a while to code this together, where the actual testing took like 80% of that time. :smile: 

Regarding harness type resources they have much (or all?) of the testing logic in each individual repo, have they not? Each individual repo knows what files it must add to get coverage and so on. AppVeyor.psm1 just kicks of the harness test function. So my thought was that the Container module could be imported and used the same was as the default-type does in ApopVeyor.psm1 in this repo. For the next iteration where one should be able to decorate each test with which container the test should run in, that need to move to a module as well I think. 

Fixed all the review comments but one, I will get to that one during the week. :smile: 

---

Review status: 3 of 8 files reviewed at latest revision, 25 unresolved discussions, some commit checks failed.

---

*[AppVeyor.psm1, line 127 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4SUVUxszok8avm11By:-L4VwOk9_xdPzaqqAnrk:b-896fix) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/AppVeyor.psm1#L127)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

Some minor grammatical tweaks (non-blocking):
This will cause the integration tests to be run in order. First, the common tests will run, followed by the unit tests. Finally the integration tests will be run in the order defined. Each integration test...
</blockquote></details>

Done.

---

*[AppVeyor.psm1, line 131 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4SUNgpOTMUuA0Y4IrZ:-L4VwQDQiq42I85w-WFS:b-896fix) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/AppVeyor.psm1#L131)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

Might be worth making it clear what will happen if an integration test is not decorated (e.g. does it run first or last?)
</blockquote></details>

Done.

---

*[AppVeyor.psm1, line 382 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4SVWVu_Ex62NdJRCjB:-L4VwR5he2f92VRX9zvz:b-896fix) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/AppVeyor.psm1#L382)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

Thirdly -> Finally
</blockquote></details>

Done.

---

*[AppVeyor.psm1, line 412 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4T6LLx2KdpVaOumZ0N:-L4VwcEdRORO4BaKh03b:b1weq8b) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/AppVeyor.psm1#L412)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

> microsoft/windowsservercore
Is it worth specifying a tag? E.g. 'microsoft/windowsservercore:latest'

Otherwise it might quite happily use old/out-of-date docker image versions if they exist on the host (I'm often guilty of not updating my docker images :grin: ). 

Non-blocking BTW - just a thought.
</blockquote></details>

Done. This will fixed differently in the next iteration where one should be able to specify which image to use, nano, servercore or any other container that suits the test. 
For now I removed it and added a defult value for that parameter as per other review comment. :smile:

---

*[AppVeyor.psm1, line 420 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4SXxgPqtfjlpy83L6H:-L4VwxP1r1RJJQhUSoX2:b-896fix) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/AppVeyor.psm1#L420)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

'The rest of the files is left' -> 'The rest of the files are left'
</blockquote></details>

Done.

---

*[AppVeyor.psm1, line 476 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4SZ1J3HDhEXq_ik381:-L4VwytUt-htkxSeQ6zx:b-o6ugvd) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/AppVeyor.psm1#L476)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

Is it possible that the container won't start for whatever reason?
</blockquote></details>

Done. Yes and no, added tests to explain, and maybe we need to change that behavior in the future if needed.

---

*[AppVeyor.psm1, line 535 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4SZQlJsitx0ANEzy-G:-L4VxFH1V4lo6eLmMfve:b-896fix) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/AppVeyor.psm1#L535)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

Blank line before this line?
</blockquote></details>

Done.

---

*[README.md, line 399 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4SbvwfvozBfiwFHo5x:-L4VxGr7GnZZnc3k5AX8:b-896fix) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/README.md#L399)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

non-blocking nitpick: 'outputted on the console` -> `sent to the console`
</blockquote></details>

Done.

---

*[README.md, line 409 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4ScINzZhjCt8RtATTj:-L4VxIc0XLMbwpn65h9c:b-896fix) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/README.md#L409)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

Non-blocking grammatical suggestion: 'This to hopefully be able to see more detailed error of what happened' -> 'This is intended to enable a more detailed error of what happened to be displayed'
</blockquote></details>

Done.

---

*[README.md, line 411 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4ScfHe4Z-FAoqa7SWN:-L4VxLweUU4YLo8qS7EG:b-896fix) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/README.md#L411)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

Non-blocking grammar: 'From the Docker log an error record is searched, and if found and an error will be thrown stopping the tests in the build worker' -> 'The Docker log will be searched for any error records. If any are found then an exception will be thrown which will stop the the tests in the build worker'. 
</blockquote></details>

Done.

---

*[README.md, line 425 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4SdTSc8CAAd9t5r4E5:-L4VxNIE5__2Vucip3zn:b-896fix) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/README.md#L425)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

Non-blocking grammar: `This is intended to enable a more detailed view of the error.`
</blockquote></details>

Done.

---

*[README.md, line 427 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4SdEcIFtCsPs4G8YrU:-L4VxOQm_D46TDlbOHxV:b-896fix) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/README.md#L427)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

Non-blocking grammar: If the container exits with exit code 0 then the Docker log will _not_ be uploaded.
</blockquote></details>

Done.

---

*[TestHelper.psm1, line 1456 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4SeZmda8jyiX8LOUDx:-L4VxPXG00vWfwbce67r:b-nv26d9) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/TestHelper.psm1#L1456)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

Nitpick: Could we use back-tick on all lines or on no lines for consistency? My preference in this case is no back-tick.
</blockquote></details>

Done. Totally missed those. Agree it's better without.

---

*[DscResource.Container/DscResource.Container.psm1, line 9 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4T-OBQb26uLzusFIBH:-L4VxVnxvnup6ILvRIcU:b-896fix) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/DscResource.Container/DscResource.Container.psm1#L9)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

Could we clarify that this is intended to be executed within the container itself? E.g.

This is the function that is executed inside the container to run the tests.
</blockquote></details>

Done.

---

*[DscResource.Container/DscResource.Container.psm1, line 101 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4SiXBIr_YGE5WTFH5q:-L4VxYPAvOik1krAkE-z:b-896fix) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/DscResource.Container/DscResource.Container.psm1#L101)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

'Creates a new container a prepares it to run tests.' -> 'Creates a new container and prepares it to run tests.'
</blockquote></details>

Done.

---

*[DscResource.Container/DscResource.Container.psm1, line 135 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4SihDxvog3fBLhT9hM:-L4VxaBu4wpIZ9VoNgCX:b8gm8w0) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/DscResource.Container/DscResource.Container.psm1#L135)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

Non-blocking nitpick: Could we default this value to simplify calls? In general I'd expect we'll always use `windowsservercore:latest` - is that correct?
</blockquote></details>

Done. Good call. This is probably the image that will be used mostly.

---

*[DscResource.Container/DscResource.Container.psm1, line 310 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4T0EWJghkLg3NdkHLT:-L4VxgBil2H7Di314ykm:b-yy2ymg) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/DscResource.Container/DscResource.Container.psm1#L310)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

stopp -> stop
</blockquote></details>

Done. Swedish again. :smile:

---

*[DscResource.Container/DscResource.Container.psm1, line 346 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4T0QkElRrFjSH9bXx_:-L4VxmBpTFMEQTRzdLGC:b-kbphr4) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/DscResource.Container/DscResource.Container.psm1#L346)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

Non-blocking: Is it worth implementing some sort of timeout here? Or perhaps this could be something looked at in a later version once we've had a chance to use this in anger a bit more.
</blockquote></details>

Yes we should have a timeout here. Good call. I will fix this during the week.

---

*[DscResource.Container/DscResource.Container.psm1, line 539 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4T1snMf242E0xpA6JM:-L4VyFKuyagRBCD7Mh6x:bb5bnwa) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/DscResource.Container/DscResource.Container.psm1#L539)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

Will this support Pester nested Context blocks? Could add it to the future TODO list if not supported as some modules have used this. Probably not worth doing now though. So non-blocking. Could add this to the .NOTES for the function though.
</blockquote></details>

Done. Added a note about this under .NOTES.

---

*[DscResource.Container/DscResource.Container.psm1, line 682 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4T33Gx-bXRE1WlbOlp:-L4VyNt-B6qyQwcf_7qy:b-896fix) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/DscResource.Container/DscResource.Container.psm1#L682)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

Should be named parameter here.
</blockquote></details>

Done.

---

*[Tests/Unit/DscResource.Container.Tests.ps1, line 153 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4T4VufT6vfMVRJCA4v:-L4VyV6c_Yh1BiehOjCv:b-896fix) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/Tests/Unit/DscResource.Container.Tests.ps1#L153)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

Nitpick: add 'without throwing'
</blockquote></details>

Done.

---

*[Tests/Unit/DscResource.Container.Tests.ps1, line 176 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4T4dScdr_UdUPT529j:-L4VyVNCMQHmgcYW2hDg:b-896fix) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/Tests/Unit/DscResource.Container.Tests.ps1#L176)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

Nitpick: add 'without throwing'
</blockquote></details>

Done.

---

*[Tests/Unit/DscResource.Container.Tests.ps1, line 335 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4T5dRIle7dWeAGRo06:-L4VyWD4iDushU65q-Wn:b-896fix) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/Tests/Unit/DscResource.Container.Tests.ps1#L335)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

"in to thinking it exists so that we can mock it."

And throughout.
</blockquote></details>

Done.

---

*[Tests/Unit/DscResource.Container.Tests.ps1, line 345 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4T4nOzw9aOHj2PuMQo:-L4VyY0uYEgDHyPKtJty:b-896fix) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/Tests/Unit/DscResource.Container.Tests.ps1#L345)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

Remove blank line
</blockquote></details>

Done.

---

*[Tests/Unit/DscResource.Container.Tests.ps1, line 550 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4T5IQrgo0jBgYfEZZe:-L4VyYyx3NqFbQ3yR9Xn:b-1b2cq1) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/Tests/Unit/DscResource.Container.Tests.ps1#L550)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

SWE :grin: stoppen -> stopped
</blockquote></details>

Done. Was spelled wrong for both Swedish and English :laughing: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-:-L4Vufnef0BDqNfPYQym:b-zhoiq7)*
<!-- Sent from Reviewable.io -->

johlju commented:
@PlagueHO added a timeout in the `Wait-Container` helper function (i kind of acknowledge the review comment by mistake).

Ready for your review again.

---

Review status: 2 of 8 files reviewed at latest revision, 15 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-:-L4fu6nj_fhfLZertNoz:b25k1tc)*
<!-- Sent from Reviewable.io -->

PlagueHO commented:
@johlju - sweet! I'll finish this up tonight for you!

PlagueHO commented:
Reviewed 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[AppVeyor.psm1, line 127 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4SUVUxszok8avm11By:-L4j4kc1sv44IOEAVpuv:b-v91zct) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/AppVeyor.psm1#L127)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Thanks for putting up with my grammatical annoyances!

---

*[AppVeyor.psm1, line 412 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4T6LLx2KdpVaOumZ0N:-L4j5KRvc0R-o98RgHu1:br30u2f) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/AppVeyor.psm1#L412)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Done. This will fixed differently in the next iteration where one should be able to specify which image to use, nano, servercore or any other container that suits the test. 
For now I removed it and added a defult value for that parameter as per other review comment. :smile:
</blockquote></details>

Cool-o! Sounds good!

---

*[AppVeyor.psm1, line 480 at r2](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4j7_2u3VInzw5rXBqZ:-L4j7_2u3VInzw5rXBq_:bxqnjyk) ([raw file](https://github.com/powershell/dscresource.tests/blob/4b6c0d74671d1341997564e4137907d13e312e91/AppVeyor.psm1#L480)):*
> ```PowerShell
> 
>                     <#
>                         This will always start the container, if for some reason
> ```

Just some minor grammatical tweaks (non-blocking nitpick):
This will always start the container. If for some reason
the container fails, the problem will be handled after waiting
for the container to finish (or fail). At that point if
the container exits with a code other than 0, then the docker
logs will be gathered and sent as an artifact. If PowerShell.exe
returned an error record then that will be thrown.

We could have waited here for X seconds to check
whether the container seems to have started the task (and not
exited with an error code). But to save seconds we
assume that the container will always be able to start the
task successfully.

---

*[DscResource.Container/DscResource.Container.psm1, line 310 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4T0EWJghkLg3NdkHLT:-L4j63jKRBrKJ6-J7bcl:bgijxs3) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/DscResource.Container/DscResource.Container.psm1#L310)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Done. Swedish again. :smile:
</blockquote></details>

Haha - I'm slowly learning Swedish words (I'm surprised I don't know more given the amount of Swedish metal I listen to :grin:)

---

*[DscResource.Container/DscResource.Container.psm1, line 517 at r2](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4j6aphxECxCkcx6S8L:-L4j6aphxECxCkcx6S8M:b-ype929) ([raw file](https://github.com/powershell/dscresource.tests/blob/4b6c0d74671d1341997564e4137907d13e312e91/DscResource.Container/DscResource.Container.psm1#L517)):*
> ```PowerShell
>         approximately one minute to list.
> 
>         Note: It will loop thru all Context-blocks, even those that was nested,
> ```

Non-blocking Nitpick: was -> were

---

*[Tests/Unit/DscResource.Container.Tests.ps1, line 550 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4T5IQrgo0jBgYfEZZe:-L4j70wB6sFVmisj5GoN:b-1qrzuc) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/Tests/Unit/DscResource.Container.Tests.ps1#L550)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Done. Was spelled wrong for both Swedish and English :laughing: 
</blockquote></details>

Hahah LOL - So I won't use Stoppen in Swedish either :grin:

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/dscresource.tests/207)*
<!-- Sent from Reviewable.io -->

johlju commented:
Think I got all of them now.

---

Review status: 6 of 8 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[AppVeyor.psm1, line 127 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4SUVUxszok8avm11By:-L4kYMlztnmAnuFNaH7M:b-xr6y0u) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/AppVeyor.psm1#L127)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

Thanks for putting up with my grammatical annoyances!
</blockquote></details>

Always! Happy that you make even the tiniest correction to the spelling/grammar. :smile:

---

*[AppVeyor.psm1, line 480 at r2](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4j7_2u3VInzw5rXBqZ:-L4kYa3p6dCg0zQKD04N:b-896fix) ([raw file](https://github.com/powershell/dscresource.tests/blob/4b6c0d74671d1341997564e4137907d13e312e91/AppVeyor.psm1#L480)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

Just some minor grammatical tweaks (non-blocking nitpick):
This will always start the container. If for some reason
the container fails, the problem will be handled after waiting
for the container to finish (or fail). At that point if
the container exits with a code other than 0, then the docker
logs will be gathered and sent as an artifact. If PowerShell.exe
returned an error record then that will be thrown.

We could have waited here for X seconds to check
whether the container seems to have started the task (and not
exited with an error code). But to save seconds we
assume that the container will always be able to start the
task successfully.
</blockquote></details>

Done.

---

*[DscResource.Container/DscResource.Container.psm1, line 310 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4T0EWJghkLg3NdkHLT:-L4kYjiKiY44CP3Otjba:b2ylc2r) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/DscResource.Container/DscResource.Container.psm1#L310)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

Haha - I'm slowly learning Swedish words (I'm surprised I don't know more given the amount of Swedish metal I listen to :grin:)
</blockquote></details>

Haha :smile: Word by word, and soon everyone will mistake you for a Swede :laughing: 

---

*[DscResource.Container/DscResource.Container.psm1, line 517 at r2](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4j6aphxECxCkcx6S8L:-L4kZ4aXEVaZdH25jPUV:b-896fix) ([raw file](https://github.com/powershell/dscresource.tests/blob/4b6c0d74671d1341997564e4137907d13e312e91/DscResource.Container/DscResource.Container.psm1#L517)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

Non-blocking Nitpick: was -> were
</blockquote></details>

Done.

---

*[Tests/Unit/DscResource.Container.Tests.ps1, line 550 at r1](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-L4T5IQrgo0jBgYfEZZe:-L4kZCyx6NqwB6z0-lbs:brjx0x5) ([raw file](https://github.com/powershell/dscresource.tests/blob/674a2966eca2015e265af23a94f6978f38c5c696/Tests/Unit/DscResource.Container.Tests.ps1#L550)):*
<details><summary><i>Previously, PlagueHO (Daniel Scott-Raynsford) wrote…</i></summary><blockquote>

Hahah LOL - So I won't use Stoppen in Swedish either :grin:
</blockquote></details>

Haha no. 'stoppad' would have been the correct word in Swedish :smile: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/dscresource.tests/207#-:-L4kZRkOhABczcPaw-bd:b-3gfxo0)*
<!-- Sent from Reviewable.io -->

kwirkykat commented:
So cool! 😄 Will merge as soon as the conflicts are resolved.

johlju commented:
@kwirkykat I have rebased. Codecoverage decreased significantly, I will look at getting that coverage up once this PR is through, if that is okay. Doing test runs now against SqlServer to verify the rebase.

@PlagueHO If you have time maybe you could sign-off on the rebase? Katie will probably merge anyway, so if you don't have time, that is okay 😄

johlju commented:
Rebase tested *successfully* here:

**SqlServerDsc (using container):** https://ci.appveyor.com/project/johlju/sqlserverdsc/build/9.0.360.0
**PSDscResources ("Default model" without container):** https://ci.appveyor.com/project/johlju/psdscresources/build/2.4.15.0
**SharePointDsc ("Harness model"):** https://ci.appveyor.com/project/johlju/sharepointdsc-p26e6/build/2.3.0.5
